### ISSUE TITLE
allow vault_generic_secret to do a more in-depth comparison of the json_data to determine changes

### ISSUE THREAD

**Issue by dansteen** [terraform-providers/terraform-provider-vault#36]
### Terraform Version
0.10.7

### Affected Resource(s)
- vault_generic_secret

### Issue
Hi!  Thanks for all the work you guys do!  I had a quick issue.

The `vault_generic_secret` (when using the `allow_read: true` config) will compare the json_data you are passing in against what's already there, and make changes as needed.  This is generally fine.  However, `vault_generic_secret` can be used to adjust anything in vault with a path.  When adjusting a non-secret path, often that path will have some settings associated with it that will always be returned, regardless of whether you set them or not.   This will cause terraform to *always* update those paths since there is technically a difference between what is set in `json_data` and what is returned.   As a concrete example, when setting `auth/aws/role/test-role` I use the following config:

```hcl
resource "vault_generic_secret" "iam-role-auth" {
  path       = "auth/aws/role/test-role"
  allow_read = true

  # 2592000 = 720h = 1m
  data_json = <<EOT
{
 "auth_type":"iam",
 "policies":"default,test-policy",
 "max_ttl":"2592000",
 "bound_iam_principal_arn":"${aws_iam_role.api.arn}"
 }
EOT
}
```
However, when querying vault, vault returns the following:
```json
{
  "allow_instance_migration":false,
  "auth_type":"iam",
  "bound_account_id":"",
  "bound_ami_id":"",
  "bound_iam_instance_profile_arn":"",
  "bound_iam_principal_arn":"xxxxx",
  "bound_iam_principal_id":"xxxxxx",
  "bound_iam_role_arn":"",
  "bound_region":"",
  "bound_subnet_id":"",
  "bound_vpc_id":"",
  "disallow_reauthentication":false,
  "inferred_aws_region":"",
  "inferred_entity_type":"",
  "max_ttl":2592000,
  "period":0,
  "policies":["default","test-policy"],
  "resolve_aws_unique_ids":true,
  "role_tag":"",
  "ttl":0
  }
```

So there are many values that are different than what I am setting (since I don't set them).  My initial thought was just to pass those default values in as well.  However, I can't do that since vault throws a number of errors as it doesn't allow many of the settings above to be set in the specific case of `auth_type: iam`.   So I'm stuck either having these policies update on every `apply` (bad) setting `allow_read: false` and just living with the drift (bad from an auditing perspective) or implementing a workaround based on env variables that sets `allow_read: true` every now and then just to make sure that there is no drift (messy).   

To avoid all this it would be great if there was an option to do a deeper comparison of the json that is returned.  Perhaps an option that allows you to decide what gets compared something like:

```compare_json: <value>```
where \<value> could be one of:
- present - only compare the json elements that are explicitly set in the provider
- all - compare all json elements regardless of whether they are explicitly set in the provider
- ?? - some other options?

I think this would greatly increase the flexibility and usability of the `generic_secret` provider for vault.

Thanks so much!

**Comment by apparentlymart** [comment]
Hi @dansteen! Thanks for this suggestion.

The `vault_generic_secret` resource was, as you noted, intended primarily for interacting with Vault's generic secret backend, rather than as a generic method to call other APIs. While Vault's API design makes it _possible_ to use for this purpose, it isn't always a great fit.

Our plan for this provider is to add specialized resources for other backends within Vault. I believe support for what you were trying to do here is, in fact, already in progress and coming soon via a set of resources specifically for the AWS auth backend. This specialized resource can then properly detect drift because it has a deeper understanding of the structure of the request and response payload for these endpoints.

**Comment by dansteen** [comment]
cool!  thanks so much!

**Comment by apparentlymart** [comment]
Hi @dansteen!

A new resource [`aws_auth_backend_role`](https://www.terraform.io/docs/providers/vault/r/aws_auth_backend_role.html) was added in #24 and included in release v1.0.0 of this provider. I think this resource should allow you to achieve what you were trying without the false-drift issues:

```hcl
resource "aws_auth_backend_role" "example" {
  backend                 = "aws"
  role                    = "test_role"
  auth_type               = "iam"
  policies                = ["default", "test-policy"]
  max_ttl                 = 2592000
  bound_iam_principal_arn = "${aws_iam_role.api.arn}"
}
```

I think this addresses your issue here, so I'm going to close this. Please let me know if anything seems amiss!
